orchagent/bfdorch.cpp

```diff
@@ -0,0 +1,412 @@
+        return false;
+    }
+    attr.id = SAI_BFD_SESSION_ATTR_SRC_IP_ADDRESS;
+    copy(attr.value.ipaddr, src_ip);
+    attrs.emplace_back(attr);
+    fvVector.emplace_back("local_addr", src_ip.to_string());
+
+    attr.id = SAI_BFD_SESSION_ATTR_DST_IP_ADDRESS;
+    copy(attr.value.ipaddr, peer_address);
+    attrs.emplace_back(attr);
+
+    attr.id = SAI_BFD_SESSION_ATTR_MIN_TX;
+    attr.value.u32 = tx_interval * 1000;
+    attrs.emplace_back(attr);
+    fvVector.emplace_back("tx_interval", to_string(tx_interval));
+
+    attr.id = SAI_BFD_SESSION_ATTR_MIN_RX;
+    attr.value.u32 = rx_interval * 1000;
+    attrs.emplace_back(attr);
+    fvVector.emplace_back("rx_interval", to_string(rx_interval));
+
+    attr.id = SAI_BFD_SESSION_ATTR_MULTIPLIER;
+    attr.value.u8 = multiplier;
+    attrs.emplace_back(attr);
+    fvVector.emplace_back("multiplier", to_string(multiplier));
+
+    if (multihop)
+    {
+        attr.id = SAI_BFD_SESSION_ATTR_MULTIHOP;
+        attr.value.booldata = true;
+        attrs.emplace_back(attr);
+        fvVector.emplace_back("multihop", "true");
+    }
+    else
+    {
+        fvVector.emplace_back("multihop", "false");
+    }
+
+    if (alias != "default")
+    {
+        attr.id = SAI_BFD_SESSION_ATTR_HW_LOOKUP_VALID;
+        attr.value.booldata = false;
+        attrs.emplace_back(attr);
+
+        Port port;
+        if (!gPortsOrch->getPort(alias, port))
+        {
+            SWSS_LOG_ERROR("Failed to locate port %s", alias.c_str());
+            return false;
+        }
+
+        attr.id = SAI_BFD_SESSION_ATTR_PORT;
+        attr.value.oid = port.m_port_id;
+        attrs.emplace_back(attr);
+
+        attr.id = SAI_BFD_SESSION_ATTR_SRC_MAC_ADDRESS;
+        memcpy(attr.value.mac, port.m_mac.getMac(), sizeof(sai_mac_t));
+        attrs.emplace_back(attr);
+
+        if (!dst_mac_provided)
+        {
+            SWSS_LOG_ERROR("Failed to create BFD session %s: destination MAC address required when hardware lookup not valid",
+                            key.c_str());
+            return false;
+        }
+
+        if (vrf_name != "default")
+        {
+            SWSS_LOG_ERROR("Failed to create BFD session %s: vrf is not supported when hardware lookup not valid",
+                            key.c_str());
+            return false;
+        }
+
+        attr.id = SAI_BFD_SESSION_ATTR_DST_MAC_ADDRESS;
+        memcpy(attr.value.mac, dst_mac.getMac(), sizeof(sai_mac_t));
+        attrs.emplace_back(attr);
+    }
+    else
+    {
+        attr.id = SAI_BFD_SESSION_ATTR_VIRTUAL_ROUTER;
+        if (vrf_name == "default")
+        {
+            attr.value.oid = gVirtualRouterId;
+        }
+        else
+        {
+            VRFOrch* vrf_orch = gDirectory.get<VRFOrch*>();
+            attr.value.oid = vrf_orch->getVRFid(vrf_name);
+        }
+
+        attrs.emplace_back(attr);
+
+        if (dst_mac_provided)
```

Comment:
I think this check can be relaxed as its ok to ignore. In future when modifying attributes are supported, we don't want to have some restriction as to remove certain attributes when some other attributes are modified.

Reply:
This check is per SAI definition that mac address is required when and only when `SAI_BFD_SESSION_ATTR_HW_LOOKUP_VALID` is `false`. Therefore, I added the check here and the opposite one for `SAI_BFD_SESSION_ATTR_HW_LOOKUP_VALID` is `true`.